### ISSUE TITLE
Add (conditional) support for UILayoutGuide

### ISSUE THREAD

**Issue by robb** [robb/Cartography#137]

**Comment by aocenas** [comment]
Hi, just wanted to ask whether it makes sense for me to try and do a PR for this. I see that you have "help wanted" tag on some issues, but not here so I am not sure if that means you want, or you are, already working on this yourself.

**Comment by nuudles** [comment]
I'd really like to get this working and am willing to help implement. The hard part is figuring out the best way to do this without ripping out a lot of the safety type checking inherent in the current iteration.

My first hope was to create a protocol like `Constrainable` and remove most instances of `View` for `Constrainable`, then extending `View` to conform to `Constrainable`, which would work, but then using the current version of Swift, we can't extend the `UILayoutSupport` protocol to conform to the `Constrainable` protocol, even if it's empty...

It sucks because even the `NSLayoutConstraint` initializer takes in `AnyObject` rather than any sort of view of protocol, so we could change the Cartography API to use `AnyObject` in all the `constrain` functions, but then you lose the type safety.

Thoughts on the best way to refactor the library to support this?

**Comment by robb** [comment]
Yeah, I feel like Cartography in its current form isn't well suited to work with `UILayoutGuide` without major refactoring. I don't have a good idea off the top of my head but I'd be happy to review any proposals.

**Comment by nuudles** [comment]
@robb So I played around with it a bit and have what may at least be a workaround for now until we can properly extend protocols with inheritance. What do you think of something like adding a method like this:

```swift
    internal func addConstraint(from: Property, to: UILayoutSupport, coefficients: Coefficients = Coefficients(), relation: NSLayoutRelation = .Equal) -> NSLayoutConstraint {
        from.view.car_translatesAutoresizingMaskIntoConstraints = false

        let layoutConstraint = NSLayoutConstraint(item: from.view,
            attribute: from.attribute,
            relatedBy: relation,
            toItem: to,
            attribute: NSLayoutAttribute.Bottom,
            multiplier: CGFloat(coefficients.multiplier),
            constant: CGFloat(coefficients.constant))

		var view = from.view
		while let superview = view.superview {
			view = superview
		}
        constraints.append(Constraint(view: view, layoutConstraint: layoutConstraint))

        return layoutConstraint
    }
```

And then adding new operators like such:

```swift
public func == <P: RelativeEquality>(lhs: P, rhs: UILayoutSupport) -> NSLayoutConstraint {
    return lhs.context.addConstraint(lhs, to: rhs)
}

public func == <P: RelativeEquality>(lhs: P, rhs: Expression<UILayoutSupport>) -> NSLayoutConstraint {
    return lhs.context.addConstraint(lhs, to: rhs.value, coefficients: rhs.coefficients[0])
}
```

This would enable you to do something like:

```swift
constrain(view) {
    (view) in
    view.top == topLayoutGuide
}
```

**Comment by aocenas** [comment]
@nuudles maybe you wanted to solve this issue: #95 ? At least I understood that this one is about ability to use `UILayoutGuide` objects directly the same way as views, for example:
```
constrain(view, myLayoutGuide) { view, guide in
    view.center == guide.center
}
```
Maybe there is some solution that would allow both to work seamlessly, but I am not sure how as top or bottomLayouteGuides can be used only on right hand side.

In case of `UILayoutGuides`, I think that something like the `Constrainable` protocol, as you mention, extending both `UIView` and `UILayoutGuide` could work. `constrain()` would take `Constrainable` instead of `UIView` and constrains would be created with anchors (or maybe not, not sure if you could just put `UILayoutGuide` into `NSLayoutConstraint.init()`, but it takes `AnyObject` so I need to check). Maybe there will be some ugly casts somewhere, for example when setting `translatesAutoresizingMaskIntoConstraints`, but otherwise I think it could be fairly clean.

**Comment by nuudles** [comment]
@aocenas You're absolutely right! My mistake. I was going on the assumption that `UILayoutGuide` was the same as `topLayoutGuide` and `bottomLayoutGuide` based on the name of those properties. Silly me. I'll move my discussion regarding `topLayoutGuide` and `bottomLayoutGuide` to issue #95 

Since `UILayoutGuide` is just a class, not a protocol, we can probably do the `Constrainable` protocol as described above and simply extend `View` and `UILayoutGuide` to conform to it. For the `translatesAutoresizingMaskIntoConstraints` I wonder if it makes sense to have another protocol to conform to or if you should just do the cast like:

```swift
if let view = from.constrainable as? View {
    view.car_translatesAutoresizingMaskIntoConstraints = false
}
```

Either way, I think that would be the only unclean aspect to the `Constrainable` approach, other than having to implement the various anchors for `UILayoutGuide` as `Edges`. Once Swift allows protocols to be extended to conform to protocols, we can even look towards having the `topLayoutGuide` and `bottomLayoutGuide` conform to `Constrainable`.
